core/modules/catalog/store/product/actions.ts

```diff
@@ -583,76 +583,58 @@ const actions: ActionTree<ProductState, RootState> = {
-      }
-      if (product.visibility === 1) { // not visible individually (https://magento.stackexchange.com/questions/171584/magento-2-table-name-for-product-visibility)
-        throw new Error(`Product query returned empty result product visibility = ${product.visibility}`)
-      }
-
-      let subloaders = []
-      if (product) {
-        const productFields = Object.keys(product).filter(fieldName => {
-          return config.entities.product.standardSystemFields.indexOf(fieldName) < 0 // don't load metadata info for standard fields
-        })
-        const attributesPromise = context.dispatch('attribute/list', { // load attributes to be shown on the product details - the request is now async
-          filterValues: config.entities.product.useDynamicAttributeLoader ? productFields : null,
-          only_visible: config.entities.product.useDynamicAttributeLoader === true,
-          only_user_defined: true,
-          includeFields: config.entities.optimize ? config.entities.attribute.includeFields : null
-        }, { root: true }) // TODO: it might be refactored to kind of: `await context.dispatch('attributes/list) - or using new Promise() .. to wait for attributes to be loaded before executing the next action. However it may decrease the performance - so for now we're just waiting with the breadcrumbs
-        if (isServer) {
-          subloaders.push(context.dispatch('setupBreadcrumbs', { product: product }))
-          subloaders.push(context.dispatch('filterUnavailableVariants', { product: product }))
-        } else {
-          attributesPromise.then(() => context.dispatch('setupBreadcrumbs', { product: product })) // if this is client's side request postpone breadcrumbs setup till attributes are loaded to avoid too-early breadcrumb switch #2469
-          context.dispatch('filterUnavailableVariants', { product: product }) // exec async
-        }
-        subloaders.push(attributesPromise)
-
-        // subloaders.push(context.dispatch('setupVariants', { product: product })) -- moved to "product/single"
-        /* if (product.type_id === 'grouped' || product.type_id === 'bundle') { -- moved to "product/single"
-          subloaders.push(context.dispatch('setupAssociated', { product: product }).then((subloaderresults) => {
-            context.dispatch('setCurrent', product) // because setup Associated can modify the product price we need to update the current product
-          }))
-        } */
-
-        context.dispatch('setProductGallery', { product: product })
-
-        if (config.products.preventConfigurableChildrenDirectAccess) {
-          subloaders.push(context.dispatch('checkConfigurableParent', { product: product }))
-        }
-      } else { // error or redirect
+    const product = await dispatch('single', { options: productSingleOptions })
+    if (product.status >= 2) {
+      throw new Error(`Product query returned empty result product status = ${product.status}`)
```

Comment:
I have a proposition in the future. What do you guys think of creating separate Error classes and define a structure for them, so the code could be shorter and more readable I think?

For example:
```TS
if (product.status >= 2) {
  throw new ProductQueryEmptyStatusError(product.status); // which will produce "Product query returned (...)"
}
if (product.visibility === 1) {
  throw new ProductQueryEmptyVisibilityError(product.visibility);
}
```